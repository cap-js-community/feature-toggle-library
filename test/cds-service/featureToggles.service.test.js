"use strict";

const cds = require("@sap/cds");
const toggles = require("../../src");

const { FEATURE, mockConfig: config } = require("../__common__/mockdata");

const server = cds.test("test/cds-test-project");

describe("cds-service", () => {
  beforeAll(async () => {
    await toggles.initializeFeatures({ config });
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  describe("endpoints", () => {
    test.each(["/rest/feature/state"])("%s GET is unauthorized", async (endpoint) => {
      const response = await server.get(endpoint, { validateStatus: () => true });
      expect(response.status).toBe(401);
    });

    test.each(["/rest/feature/redisRead", "/rest/feature/redisUpdate", "/rest/feature/redisSendCommand"])(
      "%s POST is unauthorized",
      async (endpoint) => {
        const response = await server.post(endpoint, {}, { validateStatus: () => true });
        expect(response.status).toBe(401);
      }
    );

    test.each(["/rest/feature/redisSendCommand"])("%s is forbidden for system", async (endpoint) => {
      const response = await server.post(
        endpoint,
        {},
        {
          validateStatus: () => true,
          auth: { username: "system", password: "system" },
        }
      );
      expect(response.status).toBe(403);
    });

    test("state response", async () => {
      const response = await server.get("/rest/feature/state", {
        auth: { username: "system", password: "system" },
      });
      expect(response.status).toBe(200);
      expect(response.data).toMatchSnapshot();
    });

    test("redisRead response", async () => {
      const response = await server.post(
        "/rest/feature/redisRead",
        {},
        { auth: { username: "system", password: "system" } }
      );
      expect(response.status).toBe(200);
      expect(response.data).toMatchInlineSnapshot(`{}`);
    });

    test("redisUpdate response success", async () => {
      const response = await server.post(
        "/rest/feature/redisUpdate",
        {
          key: FEATURE.A,
          value: true,
        },
        {
          auth: { username: "system", password: "system" },
        }
      );
      expect(response.status).toBe(204);
      expect(response.data).toMatchInlineSnapshot(`""`);
    });

    test("redisUpdate response validation fail", async () => {
      const response = await server.post(
        "/rest/feature/redisUpdate",
        {
          key: FEATURE.A,
          value: 100,
        },
        {
          validateStatus: () => true,
          auth: { username: "system", password: "system" },
        }
      );
      expect(response.status).toBe(422);
      expect(response.data).toMatchInlineSnapshot(`
        {
          "error": {
            "@Common.numericSeverity": 4,
            "code": "422",
            "message": "value "100" has invalid type number, must be boolean",
            "target": "test/feature_a",
          },
        }
      `);
    });
  });
});
