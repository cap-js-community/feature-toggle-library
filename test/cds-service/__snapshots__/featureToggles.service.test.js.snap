// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cds-service endpoints state 1`] = `
{
  "test/feature_a": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "boolean",
    },
    "fallbackValue": false,
  },
  "test/feature_aa": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "boolean",
      "VALIDATIONS": [
        {
          "scopes": [
            "tenant",
            "user",
          ],
        },
      ],
    },
    "fallbackValue": false,
  },
  "test/feature_b": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "number",
    },
    "fallbackValue": 1,
  },
  "test/feature_c": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "string",
    },
    "fallbackValue": "best",
  },
  "test/feature_d": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "boolean",
      "VALIDATIONS": [
        {
          "regex": "^(?:true)$",
        },
      ],
    },
    "fallbackValue": true,
  },
  "test/feature_e": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "number",
      "VALIDATIONS": [
        {
          "scopes": [
            "component",
            "layer",
            "tenant",
          ],
        },
        {
          "regex": "^\\d{1}$",
        },
      ],
    },
    "fallbackValue": 5,
  },
  "test/feature_f": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "string",
      "VALIDATIONS": [
        {
          "regex": "^(?:best|worst)$",
        },
      ],
    },
    "fallbackValue": "best",
  },
  "test/feature_g": {
    "config": {
      "ACTIVE": false,
      "SOURCE": "RUNTIME",
      "TYPE": "string",
    },
    "fallbackValue": "activeTest",
  },
  "test/feature_h": {
    "config": {
      "APP_URL": "\\.cfapps\\.sap\\.hana\\.ondemand\\.com$",
      "SOURCE": "RUNTIME",
      "TYPE": "string",
    },
    "fallbackValue": "appUrlTest",
  },
}
`;

exports[`cds-service endpoints state response 1`] = `
{
  "test/feature_a": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "boolean",
    },
    "fallbackValue": false,
  },
  "test/feature_aa": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "boolean",
      "VALIDATIONS": [
        {
          "scopes": [
            "tenant",
            "user",
          ],
        },
      ],
    },
    "fallbackValue": false,
  },
  "test/feature_b": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "number",
    },
    "fallbackValue": 1,
  },
  "test/feature_c": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "string",
    },
    "fallbackValue": "best",
  },
  "test/feature_d": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "boolean",
      "VALIDATIONS": [
        {
          "regex": "^(?:true)$",
        },
      ],
    },
    "fallbackValue": true,
  },
  "test/feature_e": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "number",
      "VALIDATIONS": [
        {
          "scopes": [
            "component",
            "layer",
            "tenant",
          ],
        },
        {
          "regex": "^\\d{1}$",
        },
      ],
    },
    "fallbackValue": 5,
  },
  "test/feature_f": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "string",
      "VALIDATIONS": [
        {
          "regex": "^(?:best|worst)$",
        },
      ],
    },
    "fallbackValue": "best",
  },
  "test/feature_g": {
    "config": {
      "ACTIVE": false,
      "SOURCE": "RUNTIME",
      "TYPE": "string",
    },
    "fallbackValue": "activeTest",
  },
  "test/feature_h": {
    "config": {
      "APP_URL": "\\.cfapps\\.sap\\.hana\\.ondemand\\.com$",
      "SOURCE": "RUNTIME",
      "TYPE": "string",
    },
    "fallbackValue": "appUrlTest",
  },
}
`;

exports[`cds-service test endpoints state 1`] = `
{
  "test/feature_a": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "boolean",
    },
    "fallbackValue": false,
  },
  "test/feature_aa": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "boolean",
      "VALIDATIONS": [
        {
          "scopes": [
            "tenant",
            "user",
          ],
        },
      ],
    },
    "fallbackValue": false,
  },
  "test/feature_b": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "number",
    },
    "fallbackValue": 1,
  },
  "test/feature_c": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "string",
    },
    "fallbackValue": "best",
  },
  "test/feature_d": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "boolean",
      "VALIDATIONS": [
        {
          "regex": "^(?:true)$",
        },
      ],
    },
    "fallbackValue": true,
  },
  "test/feature_e": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "number",
      "VALIDATIONS": [
        {
          "scopes": [
            "component",
            "layer",
            "tenant",
          ],
        },
        {
          "regex": "^\\d{1}$",
        },
      ],
    },
    "fallbackValue": 5,
  },
  "test/feature_f": {
    "config": {
      "SOURCE": "RUNTIME",
      "TYPE": "string",
      "VALIDATIONS": [
        {
          "regex": "^(?:best|worst)$",
        },
      ],
    },
    "fallbackValue": "best",
  },
  "test/feature_g": {
    "config": {
      "ACTIVE": false,
      "SOURCE": "RUNTIME",
      "TYPE": "string",
    },
    "fallbackValue": "activeTest",
  },
  "test/feature_h": {
    "config": {
      "APP_URL": "\\.cfapps\\.sap\\.hana\\.ondemand\\.com$",
      "SOURCE": "RUNTIME",
      "TYPE": "string",
    },
    "fallbackValue": "appUrlTest",
  },
}
`;
